tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogicTest.java

```diff
@@ -547,6 +558,35 @@ public void testWholeWordSearchAvailable() {
 		assertThat(findReplaceLogic.isWholeWordSearchAvailable(""), is(false));
 	}
 
+	/**
+	 * The TextViewer implementation of IFindReplaceLogic is misleading and not adhering to the
+	 * Interface: IFindReplaceTargetExtension3#replaceSelection will NOT replace the selection if
+	 * nothing was previously found. This does not generally have to be the case and thus we mock a
+	 * target that is setup like this:
+	 *
+	 * The text contained is {@code ~SELECTEDTEXT~abcd} with ~ marking the boundaries of the
+	 * selection. We perform a search for "NOTFOUND" first - the selection stays put since the
+	 * string was not found. At this point, we do not want to perform
+	 * {@code replaceSelection("NOTFOUND")} since an implementation adhering to the specification of
+	 * the function would just overwrite the current selection.
+	 */
+	@Test
+	public void onlySelectAndReplacesIfFindSuccessfulOnCustomTarget() {
+		IFindReplaceTarget mockedTarget= Mockito.mock(IFindReplaceTarget.class, withSettings().extraInterfaces(IFindReplaceTargetExtension3.class, IFindReplaceTargetExtension2.class));
+
+		when(mockedTarget.getSelectionText()).thenReturn("SELECTEDTEXT");
+		when(mockedTarget.getSelection()).thenReturn(new Point(0, "SELECTEDTEXT".length()));
+		when(mockedTarget.isEditable()).thenReturn(true);
+		when(mockedTarget.findAndSelect(anyInt(), anyString(), anyBoolean(), anyBoolean(), anyBoolean())).thenReturn(-1);
+		when(((IFindReplaceTargetExtension2) mockedTarget).validateTargetState()).thenReturn(true);
+
+		IFindReplaceLogic findReplaceLogic= new FindReplaceLogic();
+		findReplaceLogic.updateTarget(mockedTarget, true);
+		findReplaceLogic.performSelectAndReplace("NOTFOUND", "");
+
+		verify((IFindReplaceTargetExtension3) mockedTarget, times(0)).replaceSelection(anyString(), anyBoolean());
```

Comment:
Nit: It should be possible to use `none()` as a shortcut for `times(0)` 
```suggestion
		verify((IFindReplaceTargetExtension3) mockedTarget, none()).replaceSelection(anyString(), anyBoolean());
```